### PR TITLE
Network Naming scheme propagation

### PR DESCRIPTION
**Pull request by orelmisan**
### What this PR does
Before this PR:

After this PR:

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note

```

### CODE REVIEW

File: pkg/network/setup/netpod/netpod.go [hunk 1]
```diff
@@ -149,13 +147,34 @@ func WithVMIIfaceStatuses(vmiIfaceStatuses []v1.VirtualMachineInstanceNetworkInt
...
 
+	if n.dynamicPodIfaceNamingEnabled {
+		ifaceStatusesByName := vmispec.IndexInterfaceStatusByName(n.vmiIfaceStatuses, func(ifaceStatus v1.VirtualMachineInstanceNetworkInterface) bool {
+			return ifaceStatus.Name != "" && ifaceStatus.PodInterfaceName != ""
+		})
+
+		filteredNets = vmispec.FilterNetworksSpec(filteredNets, func(net v1.Network) bool {
+			_, exists := ifaceStatusesByName[net.Name]
+			return exists
+		})
```
Comment (orelmisan) [inline pkg/network/setup/netpod/netpod.go:171]: Consider adding a scenario in which the pod interface name is missing for some networks.

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/8cb9bc0bdc73452819d3dd3b25d8195cb7c10f39..66b5ac441bee90a7005e664ec2a490f338fe0042): Rebase on top of https://github.com/kubevirt/kubevirt/pull/13329.

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/277bf7ac1b8f344b6c0470aa49b02d90e222fa99..677282d3df8db97d5d7c9608b82bfa540397cbf6): Two entries in a table had the same description, causing the `pull-kubevirt-unit-test` lane to fail.

File: tests/libnet/hotplug.go [hunk 2]
```diff
@@ -123,7 +124,8 @@ func interfaceStatusFromInterfaceNames(queueCount int32, ifaceNames ...string) [
 			InterfaceName: fmt.Sprintf("eth%d", i+initialIfacesInVMI),
 			InfoSource: vmispec.NewInfoSource(
 				vmispec.InfoSourceDomain, vmispec.InfoSourceGuestAgent, vmispec.InfoSourceMultusStatus),
-			QueueCount: queueCount,
+			QueueCount:       queueCount,
```
Comment (ormergi) [inline tests/libnet/hotplug.go:127]: If this change is not mandatory please more this change to different PR.

File: pkg/virt-launcher/virtwrap/nichotplug.go [hunk 3]
```diff
@@ -151,9 +151,11 @@ func networksToHotplugWhoseInterfacesAreNotInTheDomain(vmi *v1.VirtualMachineIns
 			_, exists := indexedDomainIfaces[ifaceStatus.Name]
 			vmiSpecIface := netvmispec.LookupInterfaceByName(vmi.Spec.Domain.Devices.Interfaces, ifaceStatus.Name)
 
-			return netvmispec.ContainsInfoSource(
-				ifaceStatus.InfoSource, netvmispec.InfoSourceMultusStatus,
-			) && !exists && vmiSpecIface.State != v1.InterfaceStateAbsent && vmiSpecIface.SRIOV == nil
+			return netvmispec.ContainsInfoSource(ifaceStatus.InfoSource, netvmispec.InfoSourceMultusStatus) &&
+				ifaceStatus.PodInterfaceName != "" &&
```
Comment (ormergi) [inline pkg/virt-launcher/virtwrap/nichotplug.go:155]: Why is it necessary? 
What happen if it doesnt take in accoung podIfaceName?
Aren't infoSource=multus check good enough?

File: pkg/network/setup/podnic.go [hunk 4]
```diff
@@ -57,7 +58,12 @@ func newPhase2PodNIC(vmi *v1.VirtualMachineInstance, network *v1.Network, iface
 		return nil, err
 	}
 
-	ifaceLink, err := link.DiscoverByNetwork(podnic.handler, podnic.vmi.Spec.Networks, *podnic.vmiSpecNetwork, vmi.Status.Interfaces)
+	ifaceStatus := vmispec.LookupInterfaceStatusByName(vmi.Status.Interfaces, network.Name)
```
Comment (ormergi) [inline pkg/network/setup/podnic.go:61]: If this change is not mandatory, please do it on follow-up PR.

**Review by ormergi — COMMENTED** [review]
Thanks for the PR, please see my comment.

I find it hard to review this PR, it perform changes all over network-setup including changing some assumption on NIC discovery.
Please avoid non mandatory changes, and deffer them to follow up PRs.

Given network-setup code depends on podIfaceName coming from Multus, for all ifaces, what if Multus is not installed would it still work?

**Comment by orelmisan** [comment]
> Thanks for the PR, please see my comment.
> 
> I find it hard to review this PR, it perform changes all over network-setup including changing some assumption on NIC discovery. Please avoid non mandatory changes, and deffer them to follow up PRs.

I agree that this is not an easy PR to review, but it is expected as it does a lateral change.
I don't think that there are non-mandatory changes, but if you think that there are - please point them out.

> Given network-setup code depends on podIfaceName coming from Multus, for all ifaces, what if Multus is not installed would it still work?

The primary interface falls back to `eth0` if Multus is not installed (please see https://github.com/kubevirt/kubevirt/pull/13018), and one cannot use secondary networks without Multus.

File: pkg/network/setup/netstat.go [hunk 5]
```diff
@@ -114,45 +114,64 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
 	interfacesStatus = ifacesStatusFromGuestAgent(interfacesStatus, domain.Status.Interfaces)
 
 	primaryNetwork := netvmispec.LookupPodNetwork(vmi.Spec.Networks)
-	interfacesStatus = restorePodIfaceNameForPrimaryIface(interfacesStatus, primaryNetwork, vmi.Status.Interfaces)
+	if primaryNetwork != nil {
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netstat.go:117]: > The fact we use previous status to calculate the new status is problematic because we make the status statefull, it seems to diverge from the common convention and best practices.
I am aware this PR does not introduce this pattern, but I think it should not extend it at least.

The VMI is owned by (at least) the VMI controller and virt-handler.
When the infosource field was introduced by https://github.com/kubevirt/kubevirt/pull/6793, it caused the VMI.Status.Interfaces field to be owned by both controllers (each owning several fields).
So, they both need to keep each-other's data when calculating their own status.

> It seems update-status should retain pod-interface-name, can you tell where is it being overridden?

Each controller builds its own info and preserves the other's data (this PR extends this pattern).

File: pkg/network/setup/netpod/netpod.go [hunk 6]
```diff
@@ -106,6 +107,13 @@ func NewNetPod(vmiNetworks []v1.Network, vmiIfaces []v1.Interface, vmiUID string
 	for _, opt := range opts {
 		opt(&n)
 	}
+
+	ifaceStatusesByName := vmispec.IndexInterfaceStatusByName(n.vmiIfaceStatuses, func(ifaceStatus v1.VirtualMachineInstanceNetworkInterface) bool {
+		return ifaceStatus.Name != ""
+	})
+
+	n.podIfaceNamesByNetName = namescheme.CreateFromIfaceStatuses(n.vmiSpecNets, ifaceStatusesByName)
+
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netpod/netpod.go]: > I think introducing some "withPodIfaceNameScheme" option funtion would fit better.

This is an interesting idea, could you please elaborate what advantages is has over the proposed solution?

> Please note this change seem to make the controller reconciliation depend on a status.

It makes the virt-handler and virt-launcher depend on information generated by the VMI controller and propagated via the VMI.Status, as was proposed in https://github.com/kubevirt/community/pull/324.

File: tests/libnet/hotplug.go (hunk 2 above)
Comment (orelmisan) [reply tests/libnet/hotplug.go:127]: Without this change, the hotplug tests will fail.

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 3 above)
Comment (orelmisan) [reply pkg/virt-launcher/virtwrap/nichotplug.go:155]: On a cluster that is in a steady state, the Multus infosource and podInterfaceName should be updated together.
The added condition makes sure that they are both present.

An edge case I could think of is a hotplug scenario that was performed in close proximity to a cluster upgrade - where the Multus infosource is present, but the podInterfaceName isn't.

File: pkg/network/setup/netstat.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/network/setup/netstat.go:117]: Looking at this again, it seems I got it wrong, its not actually looks at previous iteration status, it "restores" the podIfaceName attribute that overriden by some other parts of UpdateStatus.

The "restoring" approach is confusing, and it make the code not consistent (build a iface list, override attributes, "restore" attributes..)  its hard to follow.
As mentioned above the `ifacesStatusFromMultus` seem better fit for populating the podIfaceName in this context.

[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netstat.go:117]: `ifacesStatusFromMultus` does basically the same thing.

[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netstat.go:117]: > As mentioned above the ifacesStatusFromMultus seem better fit for populating the podIfaceName in this context.

`ifacesStatusFromMultus` basically does the same thing.

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 3 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/nichotplug.go:155]: Sorry I dont follow why podIfaceName is a must for this check, could you elaborate on that?

File: pkg/network/setup/podnic.go (hunk 4 above)
Comment (orelmisan) [reply pkg/network/setup/podnic.go:61]: > It seem the current implementation already consume podIfaceName 7578195

Yes, for the primary interface.
This PR builds on the same mechanism for all interfaces.

> Question, does this change compatible with ordinal iface names (eth1,eth2)?

Yes, as VMI controller is aware of the naming scheme and propagates it to virt-handler and virt-launcher.

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 3 above)
Comment (orelmisan) [reply pkg/virt-launcher/virtwrap/nichotplug.go:155]: Lets assume one upgrades from v1.4 to v1.5.
v1.4 does not have `PodInterfaceName` populated for any interface (unless you enable the dynamic naming FG and then it will be populated for the primary interface).
So new components should assume `PodInterfaceName` could be unknown.

File: pkg/network/setup/netpod/netpod.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/network/setup/netpod/netpod.go]: > This is an interesting idea, could you please elaborate what advantages is has over the proposed solution?

Current solution blots the factory adding complexity to it https://github.com/kubevirt/kubevirt/blob/677282d3df8db97d5d7c9608b82bfa540397cbf6/pkg/network/setup/netpod/netpod.go#L89-L124.
Alternatively, we could filter ifaces/nets who's `podIfaceName` is reported at the caller.
And pass the name-scheme using an option function.

[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netpod/netpod.go]: > Current solution blots the factory adding complexity to it

> Alternatively, we could filter ifaces/nets who's podIfaceName is reported at the caller. 
And pass the name-scheme using an option function.

Some one, at some level will have to do it - and it will add a bit of filtering logic to that level.
The callers are:
1. `VirtualMachineController.vmUpdateHelperDefault` and `VirtualMachineController.vmUpdateHelperMigrationTarget`.
2. `VirtualMachineController.setupNetwork`.
3. `NetConf.Setup`.

Which level do you suggest should have the filtering logic?

File: pkg/virt-handler/non-root.go [hunk 7]
```diff
@@ -250,9 +153,6 @@ func (d *VirtualMachineController) nonRootSetup(origVMI, vmi *v1.VirtualMachineI
 	if err := d.prepareStorage(origVMI, res); err != nil {
 		return err
 	}
-	if err := d.prepareTap(origVMI, res); err != nil {
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/virt-handler/non-root.go:253]: This commit is from https://github.com/kubevirt/kubevirt/pull/13329, which this PR depends on.
The reason for it is that it contained code related to dynamic primary NIC naming.

File: pkg/network/setup/netstat.go [hunk 8]
```diff
@@ -114,45 +114,64 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
...
-		return slices.Insert(interfacesStatus, 0, v1.VirtualMachineInstanceNetworkInterface{
-			Name:             prevPrimaryIfaceStatus.Name,
-			PodInterfaceName: prevPrimaryIfaceStatus.PodInterfaceName,
+	return slices.Insert(interfacesStatus, 0, v1.VirtualMachineInstanceNetworkInterface{
+		Name: prevPrimaryIfaceStatus.Name,
+	})
+}
+
+// restorePodIfaceNames restores the PodInterfaceName based on the VMI controller's last report
+func restorePodIfaceNames(
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netstat.go:156]: I would appreciate actionable feedback that will make it more understandable.

File: pkg/network/setup/netpod/netpod.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/network/setup/netpod/netpod.go]: I am aware we cannot avoid the additional filtering logic, my suggestion is moving it out side the factory (NewNetPod) to keep it simple as it is now.
The only place make sense to me to do these logic is the caller of the factory, that is NetConf.Setup.

The problem I see here is the additional code assume the WithVMIIfaceStatuses option is given, in practice it does, but when you look at the factory code its not straightforward to realize where n.vmiIfaceStatuses comes from.

Since its about cosmetic and not actual functionality issue, lets wait to see what over reviewers think.

File: pkg/network/setup/netstat.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/network/setup/netstat.go:117]: Sorry I dont understand what you mean by that, `ifacesStatusFromMultus` doesnt touch `podIfaceName`.
I am asking for changing `ifacesStatusFromMultus` to populate the `podIfaceName`, if possible.
That way we can avoid the restore approach I find confusing.

File: pkg/network/vmicontroller/status_test.go [hunk 9]
```diff
@@ -278,7 +278,7 @@ var _ = Describe("Status Update", func() {
 	})
 
 	DescribeTable("Should generate interface status for secondary interface (not matched on status) when Multus network status is reported",
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/vmicontroller/status_test.go:280]: It means that the VMI spec Interface does not have a matching interface status on VMI.Status.Interfaces.
> is it still relevant?

Yes.

File: pkg/network/setup/netstat.go [hunk 10]
```diff
@@ -114,45 +114,64 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
...
+		interfacesStatus = restorePrimaryIfaceStatus(interfacesStatus, vmi.Status.Interfaces, primaryNetwork.Name)
+	}
 	primaryInterfaceStatus, interfacesStatus := netvmispec.PopInterfaceByNetwork(interfacesStatus, primaryNetwork)
 	if primaryInterfaceStatus != nil {
 		interfacesStatus = append([]v1.VirtualMachineInstanceNetworkInterface{*primaryInterfaceStatus}, interfacesStatus...)
 	}
 
 	interfacesStatus = ifacesStatusFromMultus(interfacesStatus, multusStatusNetworksByName, vmiInterfacesSpecByName)
 
+	interfacesStatus = restorePodIfaceNames(interfacesStatus, vmi.Status.Interfaces)
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netstat.go:127]: The `NetStat.UpdateStatus` method builds the VMI.Status.Interfaces slice in stages.
Each stage enriches the previous stage - it could modify existing elements or add new ones.
When a slice grows, its underlying array could be replaced - thus there is a need to assign `interfacesStatus` multiple times.
For additional details, please see https://go.dev/blog/slices-intro and https://go.dev/ref/spec#Appending_and_copying_slices.

File: pkg/network/setup/netstat.go [hunk 11]
```diff
@@ -114,45 +114,64 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
...
+		Name: prevPrimaryIfaceStatus.Name,
+	})
+}
+
+// restorePodIfaceNames restores the PodInterfaceName based on the VMI controller's last report
+func restorePodIfaceNames(
+	interfacesStatus []v1.VirtualMachineInstanceNetworkInterface,
+	prevIfaceStatuses []v1.VirtualMachineInstanceNetworkInterface,
+) []v1.VirtualMachineInstanceNetworkInterface {
+	updatedStatuses := slices.Clone(interfacesStatus)
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netstat.go:160]: The reason to clone the slice is to work on a copy, thus assuring that the input `interfacesStatus` slice will not be mutated.

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/677282d3df8db97d5d7c9608b82bfa540397cbf6..ac397b28fef9c9b8fde119de2bf08ca885d4dac0): Rebase.

File: pkg/network/setup/netstat.go [hunk 12]
```diff
@@ -114,45 +114,64 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
...
+
+	prevIfaceStatusesByName := netvmispec.IndexInterfaceStatusByName(prevIfaceStatuses,
+		func(ifaceStatus v1.VirtualMachineInstanceNetworkInterface) bool {
+			return ifaceStatus.Name != ""
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netstat.go]: Done.

File: pkg/network/setup/netstat.go [hunk 13]
```diff
@@ -114,45 +114,64 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
...
 		})
+
+	for i, ifaceStatus := range updatedStatuses {
+		if ifaceStatus.Name == "" {
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netstat.go]: Good idea!
Done.

File: pkg/network/setup/netstat_test.go [hunk 14]
```diff
@@ -777,6 +777,23 @@ var _ = Describe("netstat", func() {
...
+		const (
+			secondaryNetworkName  = "secondary"
+			secondaryPodIfaceName = "pod123456"
+		)
+		setup.Vmi.Spec.Domain.Devices.Interfaces = append(setup.Vmi.Spec.Domain.Devices.Interfaces, newVMISpecIfaceWithBridgeBinding(secondaryNetworkName))
+		setup.Vmi.Spec.Networks = append(setup.Vmi.Spec.Networks, newVMISpecMultusNetwork(secondaryNetworkName))
+		setup.Vmi.Status.Interfaces = []v1.VirtualMachineInstanceNetworkInterface{
+			{Name: secondaryNetworkName, PodInterfaceName: secondaryPodIfaceName, InfoSource: netvmispec.InfoSourceMultusStatus},
+		}
+		Expect(setup.NetStat.UpdateStatus(setup.Vmi, setup.Domain)).To(Succeed())
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netstat_test.go:790]: The logic in `NetStat.Update` only preserves the `PodInterfaceName` reported by VMI controller, it doesn't consume it and shouldn't modify it (similarly to the Multus status info source) - so the state of the domain should not play a role IMO.

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/ac397b28fef9c9b8fde119de2bf08ca885d4dac0..77996cd9a8150120a7d8d31881be7249cbc7913b): Addressed @nirdothan's [comment](https://github.com/kubevirt/kubevirt/pull/13265#discussion_r1873357104).

File: pkg/network/setup/network.go [hunk 15]
```diff
@@ -80,6 +80,17 @@ func (v VMNetworkConfigurator) getPhase2NICs(domain *api.Domain, networks []v1.N
...
+		func(ifaceStatus v1.VirtualMachineInstanceNetworkInterface) bool {
+			return vmispec.HasIfaceStatusOriginatedFromSpec(ifaceStatus) && ifaceStatus.PodInterfaceName != ""
+		},
+	)
+
+	networks = vmispec.FilterNetworksSpec(networks, func(network v1.Network) bool {
+		_, hasPodIfaceName := ifaceStatusesByName[network.Name]
+		return hasPodIfaceName
+	})
+
```
Comment (ormergi) [inline pkg/network/setup/network.go:93]: Since there is already filtering logic on networks phase2, please consider embedding this logic in the existing logic (getPhase2NICs / newPhase2PodNIC).

File: pkg/network/setup/podnic.go [hunk 16]
```diff
@@ -57,7 +58,12 @@ func newPhase2PodNIC(vmi *v1.VirtualMachineInstance, network *v1.Network, iface
...
+	if ifaceStatus == nil {
+		return nil, fmt.Errorf("could not find interface status for %s", network.Name)
+	}
+
+	ifaceLink, err := podnic.handler.LinkByName(ifaceStatus.PodInterfaceName)
```
Comment (ormergi) [inline pkg/network/setup/podnic.go:66]: Previously `newPhase2PodNIC` would failed on **non** LinkNotFound errors to allow a state where the hotplugged network's interface is not present in the pod yet https://github.com/kubevirt/kubevirt/commit/1c5cf683a1dbf6a8cb1432fbc40c957b6d1f3e9b.

Dont we need preserve that logic to avoid the VM from failing in above case?

File: pkg/network/setup/netpod/netpod.go [hunk 17]
```diff
@@ -106,6 +107,11 @@ func NewNetPod(vmiNetworks []v1.Network, vmiIfaces []v1.Interface, vmiUID string
 	for _, opt := range opts {
 		opt(&n)
 	}
+
+	ifaceStatusesByName := vmispec.IndexInterfaceStatusByName(n.vmiIfaceStatuses, vmispec.HasIfaceStatusOriginatedFromSpec)
+
+	n.podIfaceNamesByNetName = namescheme.CreateFromIfaceStatuses(n.vmiSpecNets, ifaceStatusesByName)
```
Comment (ormergi) [inline pkg/network/setup/netpod/netpod.go]: I commented about it previously, I would move this logic to the caller keeping this factory simple as possible.
Pass the crafted  namescheme from ifaces status via opt. function.

**Review by ormergi — COMMENTED** [review]
Follow-up review:
Please see my inline comments, I am concerned about the following topics:
- network-setup phase2 fail in case pod-interface not found.
 We may need to preserve current logic to not fail on VMs who's hotplug network interface is not present yet in the pod.
- virt-handler's update-status, I find the restore logic hard to follow and maintain I hope we can avoid it. see my suggestion.
- Testing that after upgrade VM wont get into fail state, Kubeirt components doest get into reenqueues storm and hotplug / unplug work as expected.
   Did you verified the above scenarios?
   We have upgrade lane but I am not sure it covers all scenarios.

File: pkg/network/setup/netstat.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netstat.go:117]: > The update-status logic is already complex and had to follow due the fact interface status is composed from multiple sources. This change extends the interface status composition from multus source. But it seem to contradict with other sources (GA?), forcing the restore pattern.

There are several distinct phases:
1. Calculating the interface status from:
- Domain
- Host devices (SR-IOV)
- Pod cache
- Guest Agent
2. Restoration of primary interface status (if VMI controller had reported it)
3. Restoration of Multus status info source
4. Restoration of PodInterfaceNames

Steps 2-4 are restoring information set by VMI controller (multus infosource and PodInterfaceName).

> I find having the podIfaceName restore logic hard to follow in this context (which is already complex), I hope we can avoid it. How about populating the podIfaceName once, after population the iface status from other sources? (by `ifacesStatusFromMultus`).

This is what the code does.

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 3 above)
Comment (orelmisan) [reply pkg/virt-launcher/virtwrap/nichotplug.go:155]: Old VMs (prior to 1.5) will not have this condition in their virt-launcher's source code as they would run on a previous version.

File: pkg/network/setup/network.go (hunk 15 above)
Comment (orelmisan) [reply pkg/network/setup/network.go:93]: Please see the commit whose title is `virt-launcher, net setup: Skip networks with unknown PodInterfaceName`.

File: pkg/network/setup/podnic.go (hunk 16 above)
Comment (orelmisan) [reply pkg/network/setup/podnic.go:66]: I think your suggestion makes sense.
@EdDev WDYT?

File: pkg/network/setup/netpod/netpod.go (hunk 17 above)
Comment (orelmisan) [reply pkg/network/setup/netpod/netpod.go]: So I suggest we will continue the discussion on the previous [thread](https://github.com/kubevirt/kubevirt/pull/13265#discussion_r1869482040).

File: pkg/network/setup/netpod/netpod.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netpod/netpod.go]: > The only place make sense to me to do these logic is the caller of the factory, that is NetConf.Setup.

Could you please elaborate what will be the advantages of doing so?

> The problem I see here is the additional code assume the WithVMIIfaceStatuses option is given, in practice it does, but when you look at the factory code its not straightforward to realize where n.vmiIfaceStatuses comes from.

The alternative that I currently have in mind is to make `NewNetPod` take another mandatory parameter. Because it already takes seven mandatory parameters and one variadic - I find it unreasonable.

**Comment by orelmisan** [comment]
> Follow-up review: Please see my inline comments, I am concerned about the following topics:
> 
> * network-setup phase2 fail in case pod-interface not found.
>   We may need to preserve current logic to not fail on VMs who's hotplug network interface is not present yet in the pod.
> * virt-handler's update-status, I find the restore logic hard to follow and maintain I hope we can avoid it. see my suggestion.
> * Testing that after upgrade VM wont get into fail state, Kubeirt components doest get into reenqueues storm and hotplug / unplug work as expected.
>   Did you verified the above scenarios?
>   We have upgrade lane but I am not sure it covers all scenarios.

I am mostly concerned about the upgrade scenario on which the VMI controller hasn't updated the podInterfaceNames and the migration controller has handed the VMI to the target virt-handler.
We might have to add logic to the migration controller as well.

File: pkg/network/setup/podnic.go (hunk 16 above)
Comment (EdDev) [reply pkg/network/setup/podnic.go:66]: I would be worried that this is not raising an error in the e2e or unit tests.
In the setup in which we do not perform in-place hotplug, it is expected for the interface not to appear in the pod.

Before you fix anything, I would go and check what happens in that scenario and why we do not see it manifested.

File: pkg/network/setup/netstat.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netstat.go:127]: > I'm OK with your response in the context of this PR, and you can consider it resolved. I am not however OK with the was this function works, in so called "layers". I am suggesting to refactor this method in a future PR and design it in a "map/reduce" style. It collects information about interfaces from multiple sources, and in each step it both collects, and decides what to use and what to override. I am suggesting to split it into 2 separate steps: first collect all the information from all sources into a single "table" in memory, and then as a separate step, analyze the collected data and yield a consensus about which data is "right" per interface

I would be happy to review your PR if you would like to propose a refactoring of the `UpdateStatus` method.

File: pkg/network/setup/netstat.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/setup/netstat.go:160]: Because IMO this function should be self-contained and not depend on its caller's context.

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/77996cd9a8150120a7d8d31881be7249cbc7913b..ee5fe6194fb4bbcc5490108477cf97fe7f935e82): Rebase.

File: pkg/network/setup/netpod/netpod.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/network/setup/netpod/netpod.go]: > > I find it unreasonable
> 
> I find it reasonable

You need to back this opinion with something.

`NetPod` needs a lot of data in order to operate.
Depending on the scenario, the data may exist or not, but it needs to work either way.

If someone has a suggestion on how to make that object better, please send a PR so we can all take advantage of it.

--

Now, I should not comment on this here, but this may assist someone that wants to follow up on this in a separate PR: As a basic programming concept, a function, even if it is just a factory, should have a reasonable amount of variables, which is probably much less than 7. The building pattern solves this issue nicely, but it surely is not the only option. Other options include splitting the object into many, separating functionality and re-arranging the pieces.

Lets continue this on a different PR, not here.

File: pkg/network/setup/network.go (hunk 15 above)
Comment (ormergi) [reply pkg/network/setup/network.go:93]: > Please see the commit whose title is `virt-launcher, net setup: Skip networks with unknown PodInterfaceName`.

@orelmisan I am not sure what do I miss..
The filtering is done at a higher level, I am suggesting doing that in lower layers (getPhase2NICs / newPhase2PodNIC) because it already does some similar filtering.
Doing so can help centralized the filtering deing done in phase2, having it less diverged.

Not a must for this PR.

File: pkg/network/setup/podnic.go (hunk 16 above)
Comment (ormergi) [reply pkg/network/setup/podnic.go:66]: Currently there are no negative unit-tests nor e2e for hotplug. 
On e2e, we test the positive scenario, where Multus work properly and the hotplugged interface present on time.

One way to verify this e2e is making Multus hang on attaching the hotplug interface (the CNI ADD call).
Long time ago I had to do something similar, I used this CNI https://github.com/fabiand/cni-hold-protoype 
to make Multus hang on the hotplug operation.
Basically its necessary to chain this CNI to the NAD that being used for the hotplug.

File: pkg/network/setup/podnic.go (hunk 4 above)
Comment (ormergi) [reply pkg/network/setup/podnic.go:61]: This thread can be resolved.

File: pkg/network/setup/netpod/netpod.go (hunk 17 above)
Comment (ormergi) [reply pkg/network/setup/netpod/netpod.go]: This thread can be resolved.

File: pkg/network/setup/netpod/netpod.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/network/setup/netpod/netpod.go]: Fair enough.

File: pkg/network/setup/netstat.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/network/setup/netstat.go:117]: I am confused by the "restore" wording in this function context, it make me think somthing was deleted and we need to restore it, at least for me. 
(I would call this operation add/update/populate)
Following offline discussion, I realized there is no actual restoration and the current impl. functions.

I do see some complexity added I wish we could avoid, specifically with (2) & (4), I think we can combine them.
For example: set the primary iface `podIfaceName` with default value, and then call the unified `podIfaceName` population function, having it override the default if there is relevant data from Multus.

I am not sure how much time that will take. Not a must for this PR.

**Comment by ormergi** [comment]
@orelmisan I think the last topics for my side are: 
- Ensure VM doesnt get into failing state on hotplug in case the new interface not present in the pod  https://github.com/kubevirt/kubevirt/pull/13265/commits/9b5e4c5fce3d5ed8b5b5cccaee997e16b35433d6#r1874813399
- Running VM prior to this PR changes, manage to migrate, not failing
- Running VM prior to this PR changes, manage to hotplug, not failing.

**Review by Acedus — APPROVED** [review]
Thank you.
I don't have much to add aside from what other reviewers already pointed out, upgrade with live migration update policy is particularly concerning.
